Fix appending issues

fsteeg opened the pull request:
Will resolve #82 and #80.

blackwinter commented on metafix/src/main/java/org/metafacture/metafix/Value.java:
```diff
@@ -264,6 +264,23 @@ void apply(final Hash hash, final String field, final String value) {
+            },
+            /* For an indexed representation of arrays as hashes with 1, 2, 3 etc. keys.
+             * i.e. ["a", "b", "c"] as { "1":"a", "2":"b", "3": "c" }
+             * This is what is produced by JsonDecoder and Metafix itself for arrays.
```
Where does Metafix do this?

fsteeg replied:
I meant the general behavior for repeated input fields, which become a list internally, and are emitted in that indexed form (see e.g. in [this test](https://github.com/metafacture/metafacture-fix/blob/4e26885f346b4cc3542c716661b4c3084ae12bae/metafix/src/test/java/org/metafacture/metafix/MetafixMethodTest.java#L54)).

blackwinter replied:
I see, thanks for the clarification.

blackwinter commented on metafix/src/main/java/org/metafacture/metafix/Value.java:
```diff
@@ -264,6 +264,23 @@ void apply(final Hash hash, final String field, final String value) {
+            INDEXED {
+                @Override
+                void apply(final Hash hash, final String field, final String value) {
+                    final Value newHash = newHash();
+                    newHash.asHash().put(field, new Value(value));
+                    hash.add(nextIndex(hash), newHash);
```
```suggestion
                    hash.add(nextIndex(hash), newHash(h -> h.put(field, new Value(value))));
```

blackwinter commented on metafix/src/test/java/org/metafacture/metafix/MetafixRecordTest.java:
```diff
@@ -264,6 +264,113 @@ public void add() {
+    public void addWithAppendInArray() {
+        MetafixTestHelpers.assertFix(streamReceiver, Arrays.asList(
+                "add_field('names.$append','patrick')",
+                "vacuum()"),
```
Why `vacuum()`? (Also in following tests.)

blackwinter reviewed (approved):
Aside from explicit array marking and still somewhat confusing list semantics (merging), :+1:

fsteeg commented:
Thanks for the review! Addressed your feedback in https://github.com/metafacture/metafacture-fix/pull/87/commits/adb06930f92d1dae241344dc4f83087b6e75d778.

blackwinter commented:
Test `complexAppendWithArrayOfObjects()` still fails:

```
org.mockito.exceptions.verification.VerificationInOrderFailure: 
Verification in order failure
Wanted but not invoked:
streamReceiver.startEntity("4");
-> at org.metafacture.metafix.MetafixRecordTest.lambda$complexAppendWithArrayOfObjects$31(MetafixRecordTest.java:466)
Wanted anywhere AFTER following interaction:
streamReceiver.endEntity();
```

Fix string:

```
copy_field('others', 'animals[].$append')
move_field('fictional', 'animals[].$append')
add_field('animals[].$append.animal', 'earthworm')
```

```
[Mockito] Interactions of: streamReceiver
     1. streamReceiver.startRecord("1");
      -> at org.metafacture.metafix.Metafix.endRecord(Metafix.java:146)
     2. streamReceiver.startEntity("animals[]");
      -> at org.metafacture.metafix.Metafix.lambda$emit$0(Metafix.java:164)
     3. streamReceiver.startEntity("1");
      -> at org.metafacture.metafix.Metafix.lambda$emit$0(Metafix.java:164)
     4. streamReceiver.literal("animal", "dog");
      -> at org.metafacture.metafix.Metafix.lambda$emit$0(Metafix.java:169)
     5. streamReceiver.endEntity();
      -> at org.metafacture.metafix.Metafix.lambda$emit$0(Metafix.java:166)
     6. streamReceiver.startEntity("2");
      -> at org.metafacture.metafix.Metafix.lambda$emit$0(Metafix.java:164)
     7. streamReceiver.literal("animal", "cat");
      -> at org.metafacture.metafix.Metafix.lambda$emit$0(Metafix.java:169)
     8. streamReceiver.endEntity();
      -> at org.metafacture.metafix.Metafix.lambda$emit$0(Metafix.java:166)
     9. streamReceiver.startEntity("3");
      -> at org.metafacture.metafix.Metafix.lambda$emit$0(Metafix.java:164)
     10. streamReceiver.literal(
        "animal",
        "earthworm"
    );
      -> at org.metafacture.metafix.Metafix.lambda$emit$0(Metafix.java:169)
     11. streamReceiver.endEntity();
      -> at org.metafacture.metafix.Metafix.lambda$emit$0(Metafix.java:166)
     12. streamReceiver.endEntity();
      -> at org.metafacture.metafix.Metafix.lambda$emit$0(Metafix.java:166)
     13. streamReceiver.startEntity("others");
      -> at org.metafacture.metafix.Metafix.lambda$emit$0(Metafix.java:164)
     14. streamReceiver.literal("animal", "human");
      -> at org.metafacture.metafix.Metafix.lambda$emit$0(Metafix.java:169)
     15. streamReceiver.endEntity();
      -> at org.metafacture.metafix.Metafix.lambda$emit$0(Metafix.java:166)
     16. streamReceiver.endRecord();
      -> at org.metafacture.metafix.Metafix.endRecord(Metafix.java:149)
```

Is the test wrong?

fsteeg commented:
> Is the test wrong?

No, it does not work yet for (indexed) arrays of objects (only strings).

To keep changes small and focused, I'd say we disable the test and work on that in #92.

blackwinter reviewed (approved):
> To keep changes small and focused, I'd say we disable the test and work on that in #92.

Okay.
